SwiftCompilerSources/Sources/Optimizer/FunctionPasses/DeadStoreElimination.swift

```diff
@@ -0,0 +1,267 @@
+//===--- DeadStoreElimination.swift ----------------------------------------==//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// Eliminates dead store instructions.
+///
+/// A store is dead if after the store has occurred:
+///
+/// 1. The value in memory is not read until the memory object is deallocated:
+///
+///     %1 = alloc_stack
+///     ...
+///     store %2 to %1
+///     ...               // no reads from %1
+///     dealloc_stack %1
+///
+/// 2. The value in memory is overwritten by another store before any potential read:
+///
+///     store %2 to %1
+///     ...               // no reads from %1
+///     store %3 to %1
+///
+/// In case of a partial dead store, the store is split so that some of the new
+/// individual stores can be eliminated in the next round of the optimization:
+///
+///     store %2 to %1                          // partially dead
+///     ...               // no reads from %1
+///     %3 = struct_element_addr %1, #field1
+///     store %7 to %3
+/// ->
+///     %3 = struct_extract %2, #field1
+///     %4 = struct_element_addr %1, #field1
+///     store %3 to %4                          // this store is dead now
+///     %5 = struct_extract %2, #field2
+///     %6 = struct_element_addr %1, #field2
+///     store %5 to %6
+///     ...              // no reads from %1
+///     store %7 to %3
+///
+let deadStoreElimination = FunctionPass(name: "dead-store-elimination") {
+  (function: Function, context: FunctionPassContext) in
+
+  for block in function.blocks {
+
+    // We cannot use for-in iteration here because if the store is split, the new
+    // individual stores are inserted right afterwards and they would be ignored by a for-in iteration.
+    var inst = block.instructions.first
+    while let i = inst {
+      if let store = i as? StoreInst {
+        if !context.continueWithNextSubpassRun(for: store) {
+          return
+        }
+        tryEliminate(store: store, context)
+      }
+      inst = i.next
+    }
+  }
+}
+
+private func tryEliminate(store: StoreInst, _ context: FunctionPassContext) {
+  if !store.hasValidOwnershipForDeadStoreElimination {
+    return
+  }
+
+  switch store.isDead(context) {
+    case .alive:
+      break
+    case .dead:
+      context.erase(instruction: store)
+    case .maybePartiallyDead(let subPath):
+      // Check if the a partial store would really be dead to avoid unnecessary splitting.
+      switch store.isDead(at: subPath, context) {
```

Comment:
Before doing another dead-store data flow based on a subpath, I would check an extra bit on the path-insensitive data flow state that indicates the presence of either a partial of full store on all paths.

Reply:
I wanted to keep it simple. I don't think it would make a noticeable difference in compile time